### ISSUE TITLE
v9.0.0 wishlist

### ISSUE THREAD

**Issue by zricethezav** [gitleaks/gitleaks#1274]
Hey all, thanks for all the support and contributions. I'm thinking about doing a breaking change, namely merging `protect` _into_ the `detect` command. I haven't committed totally to this change but it's something that's been in the back of my mind for a while now. 

If you have other suggestions or features you'd like to see added, throw 'em in this issue.

Thanks! 

Zach

**Comment by Devvox93** [comment]
@zricethezav I'm not sure when the v9 is planned, but I think #1059 should definitely be in there, as it currently makes using .gitleaksignore in a CI environment rather unusable.

**Comment by rgmz** [comment]
I've been meaning to respond to this for weeks. To prevent me from putting this off further, I'm going to incrementally update this comment.

Right now this isn't in any particular order and is a mix of "I feel really strongly about this" or "this is an idea I had". I will try to refine and categorize it later, as some of these are half-baked.

- #1162
    In my opinion, the current `fingerprint` implementation is brittle and trying to add everything you need to `.gitleaksignore` can be a bit of a pain-point. It would be more straightforward to "fingerprint" the secret itself and use that for `.gitleaksignore`.
- **Clarify If `keywords` Matches Path**
  I wrote this down in a note a long time ago. I forget why, and I will remove this if I can't figure it out.
- **Clarify or Rework "Stopwords"**
   In my experience using Gitleaks, as well as helping other members of the community, the concept of ["stopwords"](https://github.com/gitleaks/gitleaks#configuration) is confusing and difficult to wrap your head around. Does it match the entire secret? Is it a regular expression or a static string? Is it case sensitive? etc.
- #1212
  I don't feel particularly strongly about this, but I know it's caused issues for some people.
- **Support Multiple Allowlist Targets**
  Right now you can only chose a single `regexTarget` per allowlist (global or rule). This makes it awkward when you want to ignore multiple patterns, where one is the secret (e.g., `{{ secrets.MY_SECRET}}`) and another is the match or line (e.g., `passwordFile: ...` or `token_id: ...`). The ability to definite separate regexes for `secret` and `match` or `line` would be a nice quality-of-life improvement.
- **Support Allowlist matching `path && regex`**
  Presently, `rule.path` + `rule.regex` and `rule.allowlist.paths` + `rule.allowlist.regexes` behave differently. Unlike `rule` which matches `path && regex`, the `allowlist` will match `path || regex` which can lead to unexpected results. 
- **Make `rule.id` Mandatory and the Primary Identifier**
  Right now `rule.id` is not required. This has lead to some bugs like #1261, as well as confusing error messages like "Failed to load config error="$DESCRIPTION invalid regex secret group..." that rely on the `rule.description` (there was one about the sarif output but I can't find it atm). 
- **Support Multiple Secret Groups**
 As mentioned [in this comment](https://github.com/gitleaks/gitleaks/pull/1288#issuecomment-1806195451), I was experimenting with changing `secretGroup: n` to `secretGroups: [ n1, n2, ...]`. 
 
   There are two potential use cases for this:
      1. combing multiple pieces of input (e.g., `username: "john"\npassword: "titor" returning both `john` and `titor` rather than just `titor`) 
      2. returning the first non-empty capture group when you have mutually exclusive ones.
     
- **Support Grouping Detection by Secret**
  Rather than outputting a CSV line/JSON object per result, it would be cool to have an output that show a secret with all the locations grouped together. I imagine this would be especially beneficial for SARIF to reduce the number of alerts, but I haven't used that yet and can't say for certain.

### Non-Breaking Changes

- Add stricter validation for gitleaks.toml to prevent typos from causing silent failures (there's a PR for this but I can't find it at the moment).

**Comment by zricethezav** [comment]
> In my opinion, the current fingerprint implementation is brittle and trying to add everything you need to .gitleaksignore can be a bit of a pain-point. It would be more straightforward to "fingerprint" the secret itself and use that for .gitleaksignore.

@rgmz ahh, interesting. So if the secret is `password123` the idea would be to hash that and use that as the fingerprint. Effectively leaving location and source data out of it?

**Comment by bdwyertech** [comment]
Any chance we could source the `.gitleaksignore` via GIT if path does not exist?  I'm thinking in terms of pre-receive hooks where gitleaks is getting called from basically a git config directory, with none of the actual files, just info/objects/refs etc.

**Comment by daniel-stockhausen** [comment]
Any way to disable an inherited rule from an extended base config would be very helpful 🙂

It's great to be able to ignore/exclude secret-fingerprints, path-regex, commits and single lines (via comment), but if there is just one rule in the base config which regularly creates false positives you're pretty much forced to stop using the extension mechanism and lose the ability to be kept "up-to-date" automatically 😔

**Comment by bplaxco** [comment]
The stuff for auto config refreshing mentioned here would be nice: https://github.com/gitleaks/gitleaks/issues/1557

Also being able to stream jsonl requests to it for batch scans to avoid executing it over and over again would be handy.

```
jsonl(requests) -> gitleaks -> jsonl(responses)
```

**Comment by rgmz** [comment]
> Also being able to stream jsonl requests to it for batch scans to avoid executing it over and over again would be handy.
> 
> ```
> jsonl(requests) -> gitleaks -> jsonl(responses)
> ```

Would this stream over HTTP, socket, or something else?

**Comment by bplaxco** [comment]
> Would this stream over HTTP, socket, or something else?

Stdin/stdout if you needed anything else you could always wrap the command in something

**Comment by rgmz** [comment]
v9 should have platform-agnostic file paths and line endings.

https://github.com/gitleaks/gitleaks/issues/1641 https://github.com/gitleaks/gitleaks/issues/1565

**Comment by bufferoverflow** [comment]
https://github.com/gitleaks/gitleaks/pull/1662+ would be great to simplify usage within larger CI environments

**Comment by SimonGolms** [comment]
That the tool is also available via npm https://github.com/gitleaks/gitleaks/pull/1661

**Comment by rgmz** [comment]
> Any way to disable an inherited rule from an extended base config would be very helpful 🙂

@daniel-stockhausen Released in [v8.23.1](https://github.com/gitleaks/gitleaks/releases/tag/v8.23.1).

**Comment by bplaxco** [comment]
# Idea for "Everything is a Rule"

Here were some thoughts that came up during the multi-part rules discussion. I haven't really worked through every case, but basically the idea was make pretty much all the structs rules in the config. Here are some details around that and some field changes & definitions. And then at the bottom is a small example config using it.

I'm sure there are some issues with this idea, but wanted to get it written down. ^_^

One big logic change is that the condition field would be dropped, instead the logic would be:

A rule matches if all of these are true:

- all if its direct requirements are met (e.g. regex match, keyword check, entropy check, etc)
- if it has a `requires` list defined, then all of those match
- if it has an ignores list defined, then none of those match

For the items inside those lists, they are just rules or rule refs (more on that below) so the logic is the same for those.

## Type and field details

Here are some definitions with the go type beside it for extra clarity.

### field Rule.kind string

Kind can be used to reference the kind of thing instead of id. This way you can have multiple rules looking for the same kind of thing in different contexts, file formats, etc.

### field Rule.id string

A uniq id to the rule that doesn't indicate the kind of thing returned but is only used to identify which rule is being used. 

### interface RuleRef 

This could be either a reference to a rule by pointer or a reference to a collection of results from rules matching the rule-refspec (see below)

In the config if a string following the rule-refspec is used then that's how the RuleRef will access its findings. If the fields for a rule definition are used then RuleRef will hold a pointer to the rule loaded from that section of the config.

What I think could be cool about this is that you can still keep it flat and simple, but it unifies the logic into a single tree on a similar "rule" node type.

#### rule-refspec

- format: `<rule-field-name>:<rule-field-value>`
- fields: 
  - kind: Findings generated from a rule with that kind
  - id: Findings generated from a rule with that id
  - tag: Findings genrated  from a rule with that tag in its list of tags

### field Rule.report bool

A flag for if a rule should be included in the findings. If set to false, the rule will only be ran the first time it's referenced in a RuleRef field, and those same findings will be used for other rules referencing it too.

### field Rule.requires []RuleRef 

All of these rules must have findings for this rule to fire

### field Rule.ignores []RuleRef 

At least one of these must have findings for the rule to be ignored

### field Rule.regexes []Regexp

Mostly a helper field for use in Rules.ignores rules. The contents would be OR'd together so that `["foo", "bar"]` would become `"foo|bar"` in the regex field. If both regex and regexes are set, the loader will throw an error.

### field Rule.regexTarget string AND Rule.keywordTarget string 

Used for rules defined in requires and ignores fields to scope the checks. The default value is "secret" and the options are the same as what's in Rule.proximity. This does not apply to rules referenced via rules-refspec strings, only rules directly defined in the list.

### Rule.proximity

A string representing how close the findings for the rule must be for the rule to report. This only applies to rules with requires set. If the rule has the regex field set, the match from that is used for the other findings fields too.

Example values:

- context[:n]:  must be within n lines of each other's context (called Line in gitleaks 8).
- file: must be in the same file (default)
- commit: must be within the same commit
- match[:n]: must be within n lines of each others match line (default is must be overlapping)
- secret[:n]: must be within n lines of each others match line (default is must be overlapping)

## Example Config

```toml
# rule definitions 
[[rules]]
  id = 'cafe448e'
  # kind can be used instead of id for reference to allow multiple differnt
  # rules to to provide this kind of finding 
  kind = 'aws-iam-unique-id'
  description = 'AWS IAM Unique Identifier'
  regex = '''\b(?:A3T[A-Z0-9]|ACCA|ABIA|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z2-7]{16}\b'''
  secretGroup = 1
  entropy = 4
  keywords = [
    'aws',
  ]

  [[rules.ignores]]
    keywordTarget = "secret" # the default
    keywords = [
       'abcdef',
       'cert',
       'conf',
       'cont',
       'curl',
       'example',
       'fake',
       'get',
       'name',
       'profile',
       'resource',
       'role',
       'sample',
       'serv',
       'spec',
       'test',
       'xxxxx',
    ]

  [[rules.ignores]]
    regexTarget = 'context'
    regexes = [
      # Ignore ARNs here
      '''arn:aws:iam:''',
      # Ignore example|EXAMPLE base64 encoded
      '''(?i)aws[\w\-]{0,32}[\'\"]?\s*?[:=\(]\s*?[\'\"]?[a-z0-9\/+]*(?-i)(?:ZXhhbXBsZ|V4YW1wbG|leGFtcGxl|RVhBTVBMR|VYQU1QTE|FWEFNUExF)''',
    ]

[[rules]]
  id = 'd05b51e5'
  kind = 'aws-secret-access-key'
  description = 'AWS Secret Access Key'
  regex = '''(?:^|[^a-z0-9\/+])([a-z0-9\/+]{40})(?:[^a-z0-9\/+]|$)'''
  report = false  # Only use for another rule's requires
  secretGroup = 1
  entropy = 4
  keywords = [
    'aws',
  ]
  
  # Ignore common placeholder values 
  [[rules.ignores]]
    keywordTarget = "secret" # the default
    keywords = [
       'abcdef',
       'cert',
       'conf',
       'cont',
       'curl',
       'example',
       'fake',
       'get',
       'name',
       'profile',
       'resource',
       'role',
       'sample',
       'serv',
       'spec',
       'test',
       'xxxxx',
    ]

[[rules]]
  id = 'f01c31a2'
  description = 'AWS Key ID and Secret Key'
  proximity = 'context:5'
  requires = [
    'kind:aws-secret-access-key',
    'kind:aws-iam-unique-id',
  ]

[[ignores]]
# replaces the global allowlist
# can also reference rules from the rule definitions using a RuleRef if helpful

[[requires]]
# basically a global require list before any of the rules can match
# probably not used much, but it'd be easy to add since everything's using the same logic
# can also reference rules from the rule definitions using a RuleRef if helpful
```

# Response format:

For a single top level finding:

```json
{
  "id": "...",
  "kind": "...",
  "description": "...",
  "tags": "...",
  "context": "...(replaces line since there can be multiple lines here)...",
  "match": "...",
  "secret": "...",
  "requirements": [
  ],
}
```
Where `requirements` are the same format as the finding but those were returned for the requirements within the specified proximity.

-----

Or something like that. There's bits that seem wonky still in my head but I think something _like_ that could be interesting.

**Comment by zricethezav** [comment]
@twpayne curious if you have any thoughts on what you want to see from a v9 since https://github.com/twpayne/chezmoi uses gitleaks as a library. Also curious what the idea gitleaks API would look like for you? There is some awkwardness with the `detect` and `sources` package right now where concepts are not super clear, e.g., `detectGit` parsing git diffs or `detectFiles` reading files. In both of those cases, detectGit and detectFiles, the fragment generation should be part of the `sources` package imo.

If anyone has any other input on what their ideal gitleaks API would look like please let me know!

@bufferoverflow @bplaxco @rgmz @craigmarksmith

**Comment by twpayne** [comment]
To be honest, the current API is fine for chezmoi. I create a detector:

https://github.com/twpayne/chezmoi/blob/96b8b727683f60494daa4e44d726e1bdf7201dba/internal/cmd/config.go#L1436-L1441

and then use the `DetectBytes` method to get a list of `Finding`s:

https://github.com/twpayne/chezmoi/blob/96b8b727683f60494daa4e44d726e1bdf7201dba/internal/cmd/addcmd.go#L91-L98

Personally, I only need to scan `[]byte`s in memory, so I don't need the `Source` abstraction. I would also avoid coupling whatever generates the data to be scanned with the scanning itself. For example, it would be very bad if gitleaks could only scan files or directories on disk, instead of more generic `[]byte`s or `io.Reader`s. This also frees the application to manage the source and any concurrency, for example it might concurrently walk a directory for speed [1], or scan generated data that never touches disk.

For me, a sufficient API would be:

```go
type Detector struct {}

func NewDetector(options ...DetectorOption) (*Detector, error)

func (d *Detector) ScanBytes([]byte) []Finding

func (d *Detector) ScanReader(io.Reader) (<-chan Finding, error)
```

Once created, a `Detector` should be safe for concurrent use by multiple goroutines.

Note that I am not familiar with the internals of gitleaks, so there may be reasons I've missed that mean that this proposed API is not sufficient.

[1] See https://github.com/twpayne/find-duplicates for an example of walking a directory tree concurrently.

**Comment by DmitriyAlergant** [comment]
Have you considered using **tiktoken** tokens density calculation as as additional measure to raw character-level enthropy, as a means to reducing false positives? And hopefully allowing to lower the raw character enthropy thresholds.

Generally speaking, words and statements that have human meaning or programming language meaning use longer tokens, while high-enthropy random strings have on average 1.3 - 1.7 characters per token but most often <= 2 characters per token.

It is not a silver bullet and does have its own pitfalls, and some types of legitimate code e.g. in-line CSV stokes data has very low tokens density, although most csv data is broken by spaces into sepate words so this can be dealt with. 

Still, LLM token density calculated by tiktoken may end up having good sensitivity and specificity, and is also very fast.

**Comment by zricethezav** [comment]
v9.0.0 work will likely move to https://github.com/betterleaks/betterleaks, specifically the `v2-dev` branch
